src/protocols/der/der.h

```diff
@@ -0,0 +1,171 @@
+	[FR_TYPE_UINT32] = (bool []){[FR_DER_TAG_INTEGER] = true, [FR_DER_TAG_ENUMERATED] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_UINT64] = (bool []){[FR_DER_TAG_INTEGER] = true, [FR_DER_TAG_ENUMERATED] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_INT8] = (bool []){[FR_DER_TAG_INTEGER] = true, [FR_DER_TAG_ENUMERATED] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_INT16] = (bool []){[FR_DER_TAG_INTEGER] = true, [FR_DER_TAG_ENUMERATED] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_INT32] = (bool []){[FR_DER_TAG_INTEGER] = true, [FR_DER_TAG_ENUMERATED] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_INT64] = (bool []){[FR_DER_TAG_INTEGER] = true, [FR_DER_TAG_ENUMERATED] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_OCTETS] = (bool []){[FR_DER_TAG_BITSTRING] = true, [FR_DER_TAG_OCTETSTRING] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_STRING] = (bool []){[FR_DER_TAG_OID] = true, [FR_DER_TAG_UTF8_STRING] = true, [FR_DER_TAG_PRINTABLE_STRING] = true, [FR_DER_TAG_T61_STRING] = true, [FR_DER_TAG_IA5_STRING] = true, [FR_DER_TAG_VISIBLE_STRING] = true, [FR_DER_TAG_GENERAL_STRING] = true, [FR_DER_TAG_UNIVERSAL_STRING] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_DATE] = (bool []){[FR_DER_TAG_UTC_TIME] = true, [FR_DER_TAG_GENERALIZED_TIME] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_TLV] = (bool []){[FR_DER_TAG_SEQUENCE] = true, [FR_DER_TAG_SET] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_STRUCT] = (bool []){[FR_DER_TAG_BITSTRING] = true, [FR_DER_TAG_SEQUENCE] = true, [FR_DER_TAG_SET] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_GROUP] = (bool []){[FR_DER_TAG_SEQUENCE] = true, [FR_DER_TAG_SET] = true, [FR_DER_TAG_MAX] = false}
+};
+
+/*
+ *	Return true if the given type can be encoded as the given tag.
+ * 		@param[in] type The fr_type to check.
+ * 		@param[in] tag The der tag to check.
+ * 		@return true if the type can be encoded as the given tag.
+ */
+static inline CC_HINT(always_inline) bool fr_type_to_der_tag_valid(fr_type_t type, fr_der_tag_num_t tag)
+{
+	return fr_type_to_der_tags[type][tag];
+}
+
+static int fr_type_to_der_tag_defaults[] = {
+	[FR_TYPE_NULL] = FR_DER_TAG_NULL,
+	[FR_TYPE_BOOL] = FR_DER_TAG_BOOLEAN,
+	[FR_TYPE_UINT8] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_UINT16] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_UINT32] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_UINT64] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT8] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT16] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT32] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT64] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_OCTETS] = FR_DER_TAG_OCTETSTRING,
+	[FR_TYPE_STRING] = FR_DER_TAG_UTF8_STRING,
+	[FR_TYPE_DATE] = FR_DER_TAG_GENERALIZED_TIME,
+	[FR_TYPE_TLV] = FR_DER_TAG_SEQUENCE,
+	[FR_TYPE_STRUCT] = FR_DER_TAG_SEQUENCE,
+	[FR_TYPE_GROUP] = FR_DER_TAG_SEQUENCE
+};
+
+static inline CC_HINT(always_inline) fr_der_tag_num_t fr_type_to_der_tag_default(fr_type_t type)
+{
+	return fr_type_to_der_tag_defaults[type];
```

Comment:
what about data types like `FR_TYPE_IPADDR` which aren't listed in the above table?  The C compiler will initialize the array entry to `0`.  Is there a `FR_DER_TAG_INVALID` which is defined as `0` ?